File: src-tauri/src/system/commands.rs
use super::{fs_tree::FSDirectory, {SystemResult, SystemError}};

use std::path::PathBuf;

// ------------------
// - tauri commands -
// ------------------

#[tauri::command(rename_all = "snake_case")]
pub fn access_directory(directory: String) -> SystemResult<FSDirectory>
{
    let directory = PathBuf::from(directory.as_str());

    let mut file_tree = FSDirectory::try_from(directory)?;

    file_tree.populate();

    return Ok(file_tree);
}

#[tauri::command(rename_all = "snake_case")]
pub fn get_parent_directory(path: String) -> SystemResult<FSDirectory>
{
    let path = PathBuf::from(path.as_str());

    if path.exists()
    {
        return match path.parent()
        {
            Some(parent) => parent.to_owned().try_into(),
            None => Err(SystemError::DirectoryUnderflowError(path))
        };
    }
    else {
        return Err(SystemError::InvalidDirectoryError(path.display().to_string()));
    }
}

#[tauri::command(rename_all = "snake_case")]
pub fn resolve_environment_variable(environment_variable: String) -> Option<String>
{
    let bytes = environment_variable.as_bytes();
    
    // checking if environment variable follows the %...% format
    if bytes.len() > 2 && (bytes[0] == b'%' && bytes[bytes.len() - 1] == b'%')
    {
        // should always return some
        return match std::env::var_os(String::from_utf8_lossy(&bytes[1..bytes.len() - 1]).to_string())
        {
            Some(val) => Some(val.to_string_lossy().to_string()),
            None => None
        }
    }
    // checking if the environment variable follows the $... format
    else if bytes.len() > 1 && bytes[0] == b'$'
    {
        // should always return Some
        return match std::env::var_os(String::from_utf8_lossy(&bytes[1..]).to_string())
        {
            Some(val) => Some(val.to_string_lossy().to_string()),
            None => None
        }
    }
    // otherwise, its invalid
    else
    {
        return None
    }
}

File: src/backend/Commands.ts
import { FSDirectory } from "./FSNode";
import { invoke } from "@tauri-apps/api";
import { UserConfig } from "./UserConfig";

// -----------------------
// - UserConfig commands -
// -----------------------

export async function serialize_user_config(user_config: UserConfig): Promise<void> {
    return invoke<void>("serialize_user_config", { user_config });
}

export async function deserialize_user_config(): Promise<UserConfig> {
    return invoke<UserConfig>("deserialize_user_config");
}

export async function get_default_user_config(): Promise<UserConfig> {
    return invoke<UserConfig>("get_default_user_config");
}

// -------------------
// - System commands -
// -------------------

export async function access_directory(directory: string): Promise<FSDirectory> {
    return invoke<FSDirectory>("access_directory", { directory: directory });
}

export async function get_parent_directory(path: string): Promise<FSDirectory> {
    return invoke<FSDirectory>("get_parent_directory", { path: path });
}

export async function resolve_environment_variable(environmentVariable: string): Promise<string> {
	return invoke<string>("resolve_environment_variable", { environment_variable: environmentVariable });
}

export async function parsePath(path: string): Promise<string> {
    return invoke<string>("parse_path", { path: path });
}
